pallets/allocations/src/lib.rs

```diff
@@ -77,84 +82,99 @@ pub mod pallet {
-			// standard mathematic operands. The risk of {over, under}flow is void as this operation will
-			// effectively take a part of `amount` and thus always produce a lower number. (We use Perbill to
-			// represent percentages)
-			let amount_for_protocol = T::ProtocolFee::get() * amount;
-			let amount_for_grantee = amount.saturating_sub(amount_for_protocol);
-
-			<CoinsConsumed<T>>::put(coins_that_will_be_consumed);
+			// allocate the coins to the proxy account
+			T::Currency::resolve_creating(&T::PalletId::get().into_account(), T::Currency::issue(full_issuance));
+
+			// send to accounts, unfortunately we need to loop again
+			let mut full_protocol: BalanceOf<T> = Zero::zero();
+			for (account, amount) in batch.iter().cloned() {
+				let amount_for_protocol = T::ProtocolFee::get() * amount;
+				let amount_for_grantee = amount.saturating_sub(amount_for_protocol);
+				T::Currency::transfer(
+					&T::PalletId::get().into_account(),
+					&account,
+					amount_for_grantee,
+					ExistenceRequirement::KeepAlive,
+				)?;
+				full_protocol = full_protocol.saturating_add(amount_for_protocol);
+			}
 
-			T::Currency::resolve_creating(&T::PalletId::get().into_account(), T::Currency::issue(amount));
+			// send protocol fees
 			T::Currency::transfer(
 				&T::PalletId::get().into_account(),
 				&T::ProtocolFeeReceiver::account_id(),
-				amount_for_protocol,
-				// we use `KeepAlive` here because we want the guarantee that the funds left
-				// won't be considered dust, which would prevent us from sending the rest to
-				// the grantee.
-				ExistenceRequirement::KeepAlive,
-			)?;
-			T::Currency::transfer(
-				&T::PalletId::get().into_account(),
-				&to,
-				amount_for_grantee,
+				full_protocol,
 				ExistenceRequirement::AllowDeath,
 			)?;
 
-			Self::deposit_event(Event::NewAllocation(to, amount_for_grantee, amount_for_protocol, proof));
 			Ok(Pays::No.into())
 		}
-	}
 
-	#[pallet::event]
-	#[pallet::generate_deposit(pub(super) fn deposit_event)]
-	pub enum Event<T: Config> {
-		/// An allocation was triggered \[who, value, fee, proof\]
-		NewAllocation(T::AccountId, BalanceOf<T>, BalanceOf<T>, Vec<u8>),
+		/// Can only be called by an oracle, trigger a token mint and dispatch to
+		/// `amount`, minus protocol fees
+		#[pallet::weight(
+			<T as pallet::Config>::WeightInfo::allocate()
+		)]
+		// we add the `transactional` modifier here in the event that one of the
+		// transfers fail. the code itself should already prevent this but we add
+		// this as an additional guarantee.
+		#[transactional]
+		#[deprecated(note = "allocate is sub-optimized and chain heavy")]
+		pub fn allocate(
+			origin: OriginFor<T>,
+			to: T::AccountId,
+			amount: BalanceOf<T>,
+			_proof: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let one_batch: BoundedVec<(T::AccountId, BalanceOf<T>), T::MaxAllocs> = vec![(to, amount)]
+				.try_into()
+				.expect("one element should fit inside the bounded vec; qed");
```

Comment:
How about
```suggestion
			let one_batch: BoundedVec<(T::AccountId, BalanceOf<T>), T::MaxAllocs> = bounded_vec![(to, amount)];
```
If you take this suggestion then `use frame_support::bounded_vec;` too.